runtime/near-evm-runner/src/types.rs

```diff
@@ -96,6 +101,113 @@ pub struct MetaCallArgs {
     pub input: Vec<u8>,
 }
 
+#[derive(Debug, Eq, PartialEq)]
+pub struct EthTransaction {
+    pub nonce: U256,
+    pub gas_price: U256,
+    pub gas: U256,
+    pub to: Option<Address>,
+    pub value: U256,
+    pub data: Vec<u8>,
+}
+
+impl EthTransaction {
+    pub fn rlp_append_unsigned(&self, s: &mut RlpStream, chain_id: Option<u64>) {
+        s.begin_list(if chain_id.is_none() { 6 } else { 9 });
+        s.append(&self.nonce);
+        s.append(&self.gas_price);
+        s.append(&self.gas);
+        match self.to.as_ref() {
+            None => s.append(&""),
+            Some(address) => s.append(address),
+        };
+        s.append(&self.value);
+        s.append(&self.data);
+        if let Some(chain_id) = chain_id {
+            s.append(&chain_id);
+            s.append(&0u8);
+            s.append(&0u8);
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct EthSignedTransaction {
+    pub transaction: EthTransaction,
+    pub v: u64,
+    pub r: U256,
+    pub s: U256,
+}
+
+fn vrs_to_arr(v: u8, r: U256, s: U256) -> [u8; 65] {
+    let mut result = [0u8; 65];
+    result[0] = v;
+    r.to_big_endian(&mut result[1..33]);
+    s.to_big_endian(&mut result[33..65]);
+    result
+}
+
+impl EthSignedTransaction {
+    /// Returns sender of given signed transaction by doing ecrecover on the signature.
+    pub fn sender(&self) -> Address {
+        let mut rlp_stream = RlpStream::new();
+        // See details of CHAIN_ID computation here - https://github.com/ethereum/EIPs/blob/master/EIPS/eip-155.md#specification
+        let chain_id =
+            if [27u64, 28u64].contains(&self.v) { None } else { Some((self.v - 35) / 2) };
```

Comment:
is it somehow guaranteed that `self.v >= 35`?

Reply:
the link above defines what `v` can be: 27/28 or 35 + chain_id*2 + {0, 1}